### PR TITLE
Upgrade lumino to v2024.3.25, improve CodeMirror shortcuts handling

### PR DESCRIPTION
**Pull request by krassowski**
## References

- Fixes regressions caused by https://github.com/jupyterlab/lumino/pull/662 due to our code calling `preventDefault` to prevent default CodeMirror actions - instead we filter these out in the first place.
- Adds commands allowing to customise shortcuts for:
  - delete line (default <kbd>Ctrl</kbd> + <kbd>D</kbd> or <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>K</kbd>) and extend selection (default <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>D</kbd>):
     - this is a change in behaviour because:
        - issue [#14548](https://github.com/jupyterlab/jupyterlab/issues/14548) asking for Ctrl + D to be used for delete line (like in JupyterLab 3.x) had 17 :+1:
        - issue [#15432](https://github.com/jupyterlab/jupyterlab/issues/15432) asking for Ctrl + D to be used for extending selection had only one reporter
     - fixes https://github.com/jupyterlab/jupyterlab/issues/16096; the new behaviour is to delete the line but the shortcut can now be disabled from settings if a user prefers it to always open the ToC
     - fixes https://github.com/jupyterlab/jupyterlab/issues/14548
  -  toggle comment (default <kbd>Ctrl</kbd> + <kbd>/</kbd>; this can be used to comment multiple lines too) and toggle block comment (default <kbd>Alt</kbd> + <kbd>A</kbd>; note this does not work in Python because CodeMirror does not define block comment syntax for Python)
     - fixes https://github.com/jupyterlab/jupyterlab/issues/4778  (which has 42 :+1:)
     - closes https://github.com/jupyterlab/jupyterlab/issues/15525
     - closes https://github.com/jupyterlab/jupyterlab/issues/13128
    

## Code changes

None

## User-facing changes

image

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by krassowski** [comment]
Yup, the same tests are failing here:

```
Notice:   4 failed
    [galata] › test/galata/notebook.spec.ts:22:7 › Notebook Tests › Create Markdown cell ───────────
    [galata] › test/galata/notebook.spec.ts:40:7 › Notebook Tests › Create Raw cell ────────────────
    [galata] › test/galata/notebook.spec.ts:59:7 › Notebook Tests › Create Code cell ───────────────
    [jupyterlab] › test/jupyterlab/cells.test.ts:41:7 › Run Cells With Keyboard › Run code cell with Ctrl + Enter 
```

@fcollonval do you think this could be related to https://github.com/jupyterlab/lumino/pull/662?

**Comment by krassowski** [comment]
So this indeed is caused by https://github.com/jupyterlab/lumino/pull/662.

1. We use `preventDefault` in JupyterLab to modify CodeMirror behaviour here:

https://github.com/jupyterlab/jupyterlab/blob/41be86d5142a793ab9cb8b9eb99f92a33c8dc441/packages/codemirror/src/commands.ts#L77-L85

2. Even if we disable that login (which leads to inserting a new line when it should run a cell), CodeMirror still runs `newlineAndIndent` on Ctrl + Enter, which prevents default.

**Comment by krassowski** [comment]
I think we could filter the `defaultKeymap`, here:

https://github.com/jupyterlab/jupyterlab/blob/41be86d5142a793ab9cb8b9eb99f92a33c8dc441/packages/codemirror/src/extension.ts#L701-L714

**Comment by krassowski** [comment]
This also means that Ctrl+Shift+K no longer opens Table Of Contents when in edit mode (as it not only deletes the current line; previously it did both: https://github.com/jupyterlab/jupyterlab/issues/16096).

Given the issues:
- https://github.com/jupyterlab/jupyterlab/issues/16096
- https://github.com/jupyterlab/jupyterlab/issues/14548 (this one has 17 upvotes)

I propose that we disable these default keybindings and add proper commands for them.

File: packages/codemirror/src/extension.ts [hunk 1]
```diff
@@ -798,16 +810,15 @@ export namespace EditorExtensionRegistry {
       }),
       Object.freeze({
         name: 'extendSelection',
-        default: true,
+        default: false,
```
Comment (krassowski) [inline packages/codemirror/src/extension.ts:813]:
```suggestion
        default: true,
```

**Comment by krassowski** [comment]
It looks like the galata tests are failing because `Escape` does not leave the edit mode. I think there may be a similar CM issue with a similar solution. Also I need to change the expected shortcut for extending selection.

File: galata/test/jupyterlab/file-edit.test.ts
Comment (fcollonval) [inline galata/test/jupyterlab/file-edit.test.ts:1]: :heart:

**Review by fcollonval — APPROVED** [review]
Thanks @krassowski

**Pull request merged**